src/compiler/utilities.ts

```diff
@@ -6258,6 +6258,33 @@ namespace ts {
         return isSourceFile(node) || isModuleBlock(node) || isBlock(node) && isFunctionLike(node.parent);
     }
 
+    /**
+     * @returns Whether a node or its children refer to a different `this` than its parent.
+     * @internal
+     */
+    export function isNewThisScope(node: Node): boolean {
+        switch (node.kind) {

```

Comment:
all kinds of decorators always refer to the scope containing the class declaration:

```ts
@decorator(this)
class C {
  @decorator(this)
  prop = 1;
  @decorator(this)
  method(@decorator(this) foo: string) {}
}
```

Reply:
Added test cases 👍